[color-picker] First stab at integrating `<space-picker>`

DmitrySharabin opened the pull request:
Here is what it looks like:

<img width="1003" alt="image" src="https://github.com/user-attachments/assets/7c6bb3be-b1ae-4564-8a92-1668bbcc0364">

Live preview: https://deploy-preview-155--color-elements.netlify.app/src/color-picker/

LeaVerou commented on src/color-picker/README.md:
````diff
@@ -48,7 +48,7 @@ As with other components, you can listen to the `colorchange` event:
 ```
 
-### Dynamic
+<!-- ### Dynamic
````
This is still useful! But we may want to rework it a bit so that it's not literally replicating a space picker.

DmitrySharabin replied:
Yeah, I just commented out so we could rework it.

DmitrySharabin replied:
I still don't know what the best alternative for this is. Moreover, I have already demonstrated something similar in the Slots section. I'd be grateful for any ideas.

LeaVerou commented on src/color-picker/color-picker.css:
```diff
@@ -4,6 +4,13 @@
 }
 
+#space_picker {
+	flex: 0 1 100%; /* occupy the entire row */
```
Why?

DmitrySharabin replied:
I didn't want to change the flex layout we already use to something else or add a wrapper around the sliders and color swatch to position the space picker above the rest of the elements of the picker, so I did this. Now, I switched to the grid layout.

LeaVerou commented on src/color-picker/color-picker.css:
```diff
@@ -4,6 +4,13 @@
 
+#space_picker {
+	flex: 0 1 100%; /* occupy the entire row */
+	padding: initial;
```
???

DmitrySharabin replied:
Taking into account the following comment, I change it. This was just for alignment purposes.

LeaVerou commented on src/color-picker/color-picker.css:
```diff
@@ -4,6 +4,13 @@
+#space_picker {
+	flex: 0 1 100%; /* occupy the entire row */
+	padding: initial;
+	border: none;
```
I'd instead make it transparent and show on hover

DmitrySharabin replied:
Good idea! Thanks. Done.

LeaVerou commented on src/color-picker/color-picker.js:
```diff
@@ -1,14 +1,17 @@
+	static globalStyle = new URL("color-picker-global.css", import.meta.url);
 	static shadowStyle = true;
 	static shadowTemplate = `
+		<space-picker id="space_picker" part="space-picker" exportparts="picker: space-picker-base"></space-picker>
```
Can we put it in a slot, and listen to input events on the slot? Then people can slot in any arbitrary form element and as long as it produces `input` events, it will work.

DmitrySharabin replied:
Done!

LeaVerou commented on src/color-picker/color-picker.js:
```diff
@@ -27,14 +30,22 @@ const Self = class ColorPicker extends ColorElement {
 		super.connectedCallback?.();
 		this._el.sliders.addEventListener("input", this);
 		this._el.swatch.addEventListener("input", this);
+		this._el.space_picker.addEventListener("spacechange", this);
```
Wait, why does it fire spacechange and not input events? Does it do both? What's the difference? I thought spacechange was only for channel-picker, to distinguish what changed (though even there we may want to repurpose [`inputType`](https://developer.mozilla.org/en-US/docs/Web/API/InputEvent/inputType) for that)

DmitrySharabin replied:
> Does it do both? What's the difference?

It does. When the `<space-picker>` value is set programmatically, we can't use the `input` event to handle it. In this case, we indeed don't need it, so I switched to listening to `input` events on the slot, as you suggested. 

>(though even there we may want to repurpose [inputType](https://developer.mozilla.org/en-US/docs/Web/API/InputEvent/inputType) for that)

I tried dispatching the `input` event using the [`InputEvent` constructor](https://developer.mozilla.org/en-US/docs/Web/API/InputEvent/InputEvent) and passing the `inputType` option with the value of `spaceChange` (in the `<space-picker>` `input` event handler). Still, it seems it wasn't passed along. IDK why. 🤷‍♂️

LeaVerou commented on src/color-picker/color-picker.js:
```diff
@@ -145,6 +158,11 @@ const Self = class ColorPicker extends ColorElement {
 		},
+		spacechange: {
+			from () {
+				return this._el.space_picker;
```
Hm that will make it hard to slot custom elements, so I'd revert back to glue code

DmitrySharabin replied:
Done!

DmitrySharabin commented:
I believe I addressed all the issues you pointed out (thank you for this), and the PR is ready for another review.

There is one issue left with the dynamic props demo: see https://github.com/color-js/elements/pull/155#discussion_r1842284461

LeaVerou commented on src/color-picker/README.md:
````diff
@@ -37,6 +37,30 @@ You can use your component instead of the default color swatch:
 </color-picker>
 ```
 
+or the default space picker:
````
```suggestion
or your own form element instead of the default space picker:
```

LeaVerou commented on src/color-picker/README.md:
```diff
@@ -37,6 +37,30 @@ You can use your component instead of the default color swatch:
+<style>
+	select.horizontal {
+		writing-mode: tb;
+		field-sizing: content;
```
The horizontal select trick is nice, but not a good option for a demo, you want anything that is not relevant to the demo itself to be relatable, not to have people scratching their heads

DmitrySharabin replied:
Got it. I'll rework it. Thanks!

LeaVerou commented on src/color-picker/color-picker.js:
```diff
@@ -50,22 +59,39 @@ const Self = class ColorPicker extends ColorElement {
 	}
 
+	get spaceResolved () {
+		return this._el.space_picker.selectedSpace;
```
What would this return if we don’t have a space picker? The space picker is interchangeable UI, we shouldn't depend on it for resolving the color space.

DmitrySharabin replied:
It looks like the default `<space-picker>` is always there (even if we have some slotted content instead of it). Its value is always in sync with the `space` property. So, the resolved color space should always be correct.

<img width="879" alt="image" src="https://github.com/user-attachments/assets/ba741530-2257-4e2d-a06e-600cae41ebfb">

What am I missing?

LeaVerou replied:
Hmmmm. It still seems like a hack but ok

LeaVerou reviewed (changes requested):
Almost there, see comments!

DmitrySharabin commented on src/color-picker/README.md:
````diff
@@ -37,6 +37,30 @@ You can use your component instead of the default color swatch:
+
+```html
+<color-picker space="oklch" color="oklch(60% 30% 180)">
+	<select slot="space-picker" size="3" class="horizontal">
````
Agreed. I'll fix it

LeaVerou commented on src/color-picker/README.md:
```diff
@@ -80,6 +102,7 @@ All attributes are reactive:
 | Name | Description |
 |------|-------------|
 | (default) | The color picker's main content. Goes into the swatch. |
+| `color-space` | An element used to set the picker color space. |
```
```suggestion
| `color-space` | An element to display (and if writable, also set) the current color space. If not provided, a `<space-picker>` is used. |
```

(with `<space-picker>` being a link)

LeaVerou commented on src/color-picker/README.md:
```diff
@@ -89,12 +112,20 @@ All attributes are reactive:
+
+| Property | Type | Description |
+|----------|------|-------------|
+| `spaceResolved` | `ColorSpace` | Color space object corresponding to the `space` property. |
```
Given that in JS, `spaceResolved` is far more useful than `space`, I wonder if we should make `space` the one with the actual color space (and make it read/write), and `spaceRaw` or something the one that reflects. Is there any precedent in HTML where the object prop is the one with the "nicer" name?

LeaVerou commented on src/color-picker/README.md:
```diff
@@ -89,12 +112,20 @@ All attributes are reactive:
 |------|-------------|
-| `input` | Fired when the color changes due to user action, either with the sliders or the color swatch's input field. |
-| `change` | Fired when the color changes due to user action, either with the sliders or the color swatch's input field. |
+| `input` | Fired when the color changes due to user action, either with the space picker, sliders or the color swatch's input field. |
```
```suggestion
| `input` | Fired when the color changes due to user action, such as adjusting the sliders, entering a color in the swatch's text field, or choosing a different color space. |
```

(and same for the one below)

LeaVerou reviewed (changes requested):
This PR LGTM based on my earlier feedback, but thinking about it some more, I think we should use `spaceId` instead of `spaceRaw`. That way it always has a meaning, even when the space is set via JS and I think it would simplify a bunch of code.

Other than that, LGTM. Thanks!

DmitrySharabin commented:
> I think we should use `spaceId` instead of `spaceRaw`. That way it always has a meaning, even when the space is set via JS and I think it would simplify a bunch of code.

Does this mean we don't need to support setting `spaceId` to a space color object and expect it always to be a string?

For example, for `input[type=number]`, we can set its `value` property to a number or to a string (that can be parsed as a number). In both cases, we'll always get a string when getting `value`. However, if we try to set its value to something that can't be parsed as a number, we'll get a warning, and `value` will be equal to `""`. Do we expect to have something like that in our case?
